### PR TITLE
Generation of istio-private jobs. 

### PR DESCRIPTION
**Pull request by brian-avery**
This: 
* Implements branching for istio-private jobs using a new utility called `generate-transform-jobs`. It may be useful in the future to add diff and print functionality to this utility as well to bring our transformation job tooling in line with the prow job tooling. 
* Renames istio-private job files for consistency as well as updating of master files.
* Fixes an issue where `generate.go` doesn't check for errors
* Fixes an issue where `generate.go` hard codes paths
* Fixes an issue where we clean up public prow cluster configuration when running `make gen` but ignore private prow cluster configuration

@jacob-delgado @ericvn Please take a look. 

Part of https://github.com/istio/istio/issues/32242. @Monkeyanator I think you may be aware of other jobs that aren't automated?

### CODE REVIEW

File: prow/config/cmd/generate.go
Comment (chizhg) [inline prow/config/cmd/generate.go]: AFAICT the tool in `prow/config/cmd` is supposed to be a generic tool for generating general Prow jobs, and the tool in `prow/genjobs/` is used for transforming the generated Prow jobs to use for private repos. So can we have this change in `prow/genjobs/` instead?

**Comment by ericvn** [comment]
I'm wondering if you just change `branch` to create the new files in prow/config/istio-private_jobs. You'll still need to update the files there to support branching on a release as you have here. Then the `make gen` in a follow-on step does the `go run prow/genjobs/main.go --configs=./prow/config/istio-private_jobs` to create the new Prow jobs.

File: prow/config/cmd/generate.go
Comment (brian-avery) [reply prow/config/cmd/generate.go]: Thanks @chizhg. 

- `genjobs` is primarily for combining public jobs with private jobs (transforms) in order to produce cluster configs for istio-private jobs. 
- `prow/config/cmd/generate.go write` generates the public cluster config based on public cluster jobs. 
- `prow/config/cmd/generate.go branch` generates jobs for a new Istio branch, currently limited to only public jobs. 

This PR adds the ability to create private jobs for new Istio branches. After these are created, `make gen` can be run, which executes `genjobs` and `prow/config/cmd/generate.go write` in order to produce the cluster configs for both public and private jobs.

File: prow/config/cmd/generate.go
Comment (chizhg) [inline prow/config/cmd/generate.go]: It looks you could just use the `privateInputDir` flag value instead of hardcoding it here. The tool is now generic and is used by other projects using Prow, so it's better not adding Istio specific logic.

File: prow/config/cmd/generate.go
Comment (chizhg) [reply prow/config/cmd/generate.go]: Thanks @brian-avery .

It's still unclear to me why you need to update this tool and also import `genjobs` as a library... Is it going to generate the config files in `prow/config/istio-private_jobs`? If yes I feel it should be another tool to do this.

File: prow/config/cmd/generate.go
Comment (brian-avery) [reply prow/config/cmd/generate.go]: +1. Apologies, this is still a work in progress. Istio testing just removed the label earlier.

File: prow/config/cmd/generate.go
Comment (brian-avery) [reply prow/config/cmd/generate.go]: Yeah, this is to generate the config files in `prow/config/istio-private_jobs`. I'm fine creating separate tools to generate public and private config. Regardless, the tool will have to pull in the Configuration struct as a library as `genjobs` should probably stay for translating the cluster config. It just means we will have to call two tools in release builder when we create the config.

File: prow/config/cmd/generate.go
Comment (brian-avery) [reply prow/config/cmd/generate.go]: I wonder if we should make the jobs version of generate.go use the `inputDir` as well. This same value is currently hard-coded in master.

File: prow/generate-transform-jobs/main_test.go
Comment (chizhg) [inline prow/generate-transform-jobs/main_test.go]: Nit: the header is inconsistent with others.

**Comment by chizhg** [comment]
Do you also need to run `generate-transform-jobs` in the `make gen` target?

**Comment by brian-avery** [comment]
> Do you also need to run `generate-transform-jobs` in the `make gen` target?

No, `make gen` is to create prow cluster configs. That's taken care of by `genconfig`. This implements the same functionality that `generate.go branch` implements for normal prow job configs but for transformation configs.

File: prow/generate-transform-jobs/main_test.go
Comment (brian-avery) [reply prow/generate-transform-jobs/main_test.go]: Thanks. Let me fix that one. At some point, we should probably go through the rest of this repo and fix the other copyrights similar to https://github.com/istio/istio/pull/23958.

File: prow/cluster/jobs/istio-private/istio/istio-private.istio.master.gen.yaml
Comment (howardjohn) [inline prow/cluster/jobs/istio-private/istio/istio-private.istio.master.gen.yaml]: This doesn't seem like it should be added

**Review by howardjohn — COMMENTED** [review]
LGTM aside from one

File: prow/cluster/jobs/istio-private/istio/istio-private.istio.master.gen.yaml
Comment (brian-avery) [reply prow/cluster/jobs/istio-private/istio/istio-private.istio.master.gen.yaml]: We're currently including that in 

1.9: https://github.com/istio/test-infra/blob/master/prow/cluster/jobs/istio-private/istio/istio-private.istio.release-1.9.gen.yaml#L117

and 1.10:https://github.com/istio/test-infra/blob/master/prow/cluster/jobs/istio-private/istio/istio-private.istio.release-1.10.gen.yaml#L117

**Review by howardjohn — APPROVED** [review]

**Review by ericvn — APPROVED** [review]
LGTM

**Comment by brian-avery** [comment]
@istio/operations-maintainers Can you please give this a stamp of approval for creating the `generate-transform-jobs` utility? Test and Release will own it (and the codeowners changes are in this PR), but because it's created in the prow directory, it requires your approval. Thanks!

**Review by cjwagner — APPROVED** [review]
Approving based on https://github.com/istio/test-infra/pull/3421#issuecomment-877233446

**Pull request merged**
